src/main/java/com/github/splendor_mobile_game/websocket/handlers/reactions/GetTokens.java

```diff
@@ -101,9 +128,301 @@ public GetTokens(int connectionHashCode, UserMessage userMessage, Messenger mess
+
+            return result * (-1);
+        }
+    }
+
+    @DataClass
+    public static class DataDTO {
+        public UUID userUuid;
+        public TokensChangeDTO tokensChangeDTO;
+
+        public DataDTO(UUID userUuid, TokensChangeDTO tokensChangeDTO) {
+            this.userUuid = userUuid;
+            this.tokensChangeDTO = tokensChangeDTO;
+        }
+    }
+
+    public class ResponseData {
+        DataDTO data;
+
+        public ResponseData(DataDTO data) {
+            this.data = data;
+        }
+    }
+
     @Override
     public void react() {
-        // TODO Auto-generated method stub
+        DataDTO dataDTO = (DataDTO) userMessage.getData();
+       
+        try {
+            Map<TokenType, Integer> tokenMap = new HashMap<TokenType, Integer>();
+            tokenMap.put(TokenType.RUBY, dataDTO.tokensChangeDTO.ruby);
+            tokenMap.put(TokenType.SAPPHIRE, dataDTO.tokensChangeDTO.sapphire);
+            tokenMap.put(TokenType.EMERALD, dataDTO.tokensChangeDTO.emerald);
+            tokenMap.put(TokenType.DIAMOND, dataDTO.tokensChangeDTO.diamond);
+            tokenMap.put(TokenType.ONYX, dataDTO.tokensChangeDTO.onyx);
+
+            validateData(dataDTO, database, tokenMap);
+            
+            Room room = database.getRoomWithUser(dataDTO.userUuid);
+            User user = database.getUser(dataDTO.userUuid);
+
+            // System.out.println("ruby on table");
+            // System.out.println(room.getGame().getTokenCount(TokenType.RUBY));
+
+            // System.out.println("sapphire on table");
+            // System.out.println(room.getGame().getTokenCount(TokenType.SAPPHIRE));
+
+            // System.out.println("emerald on table");
+            // System.out.println(room.getGame().getTokenCount(TokenType.EMERALD));
+
+            // System.out.println("diamond on table");
+            // System.out.println(room.getGame().getTokenCount(TokenType.DIAMOND));
+
+            // System.out.println("onyx on table");
+            // System.out.println(room.getGame().getTokenCount(TokenType.ONYX));
+
+            // System.out.println("---------------------------------------------------");
+
+            // System.out.println("ruby on user");
+            // System.out.println(user.getTokenCount(TokenType.RUBY));
+
+            // System.out.println("sapphire on user");
+            // System.out.println(user.getTokenCount(TokenType.SAPPHIRE));
+
+            // System.out.println("emerald on user");
+            // System.out.println(user.getTokenCount(TokenType.EMERALD));
+
+            // System.out.println("diamond on user");
+            // System.out.println(user.getTokenCount(TokenType.DIAMOND));
+
+            // System.out.println("onyx on user");
+            // System.out.println(user.getTokenCount(TokenType.ONYX));
+
+
+            // System.out.println();
+            // System.out.println();
+            // System.out.println();
+
+            changeTokens(user, room, tokenMap);
```

Comment:
After approving the user's request and making appropriate changes to the game you should mark the player that he has done his action with:
```java
user.setPerformedAction(true);
```

thus making them unable to repeat this action. Currently user can GetTokens multiple times in one turn.